File: benchmark/poc/npyiter_sizesweep_bench.py
# =============================================================================
# npyiter_sizesweep_bench.py — NumPy side of the size-tier sweep (same ids as
# npyiter_sizesweep_bench.cs). Six operation families measured across four
# element-count tiers: scalar (1), 1K, 100K, 1M.
#
#   add   — np.add(a, b, out=o)             contiguous binary ufunc nditer
#   sqrt  — np.sqrt(a, out=o)               contiguous unary ufunc nditer
#   sum   — np.sum(a)                        full reduction (PyUFunc_ReduceWrapper)
#   copy  — np.positive(a, out=o)            real ufunc nditer copy (NOT np.copyto:
#                                            that is a stripped raw-array walker)
#   sadd  — np.add(a2[::2], b2[::2], out=o)  strided binary
#   bcast — np.add(a, b1, out=o)             stride-0 broadcast binary
#
# Run: python benchmark/poc/npyiter_sizesweep_bench.py
# =============================================================================
import time
import numpy as np

fails = 0


def check(ok, what):
    global fails
    if not ok:
        fails += 1
        print(f"  CORRECTNESS FAIL: {what}")


def best_ms(body, iters, warm, rounds=5):
    for _ in range(warm):
        body()
    best = float("inf")
    for _ in range(rounds):
        t0 = time.perf_counter()
        for _ in range(iters):
            body()
        best = min(best, (time.perf_counter() - t0) * 1000.0 / iters)
    return best


def row(id_, label, ms):
    if ms >= 1.0:
        val = f"{ms:10.3f} ms"
    elif ms >= 0.001:
        val = f"{ms * 1000:10.2f} us"
    else:
        val = f"{ms * 1e6:10.1f} ns"
    print(f"{id_:<12} {label:<46} {val}")


def pick(n):
    if n <= 1:
        return 400_000, 40_000
    if n <= 1_000:
        return 200_000, 25_000
    if n <= 100_000:
        return 4_000, 600
    return 400, 80


SIZES = [("1", 1), ("1K", 1_000), ("100K", 100_000), ("1M", 1_000_000)]

print(f"NumPy {np.__version__} NpyIter size-sweep")
print(f"{'id':<12} {'aspect':<46} {'per call':>13}")
print("-" * 86)

for tag, n in SIZES:
    iters, warm = pick(n)
    rounds = 7 if n >= 100_000 else 5

    a = (np.arange(n, dtype=np.float64) % 97.0) + 1.0
    b = (np.arange(n, dtype=np.float64) % 31.0) + 2.0
    o = np.empty(n, dtype=np.float64)
    b1 = np.array([3.0], dtype=np.float64)
    a2 = (np.arange(2 * n, dtype=np.float64) % 53.0) + 1.0
    b2 = (np.arange(2 * n, dtype=np.float64) % 17.0) + 1.0
    sa = a2[::2]
    sb = b2[::2]
    so = np.empty(n, dtype=np.float64)

    # add
    np.add(a, b, out=o)
    check(o[n - 1] == a[n - 1] + b[n - 1], f"add@{tag}")
    row(f"add@{tag}", f"binary add contig f64 N={n}",
        best_ms(lambda: np.add(a, b, out=o), iters, warm, rounds))

    # sqrt
    np.sqrt(a, out=o)
    check(abs(o[n - 1] - np.sqrt(a[n - 1])) < 1e-9, f"sqrt@{tag}")
    row(f"sqrt@{tag}", f"unary sqrt contig f64 N={n}",
        best_ms(lambda: np.sqrt(a, out=o), iters, warm, rounds))

    # sum
    s = np.sum(a)
    check(abs(float(s) - float(np.sum(a))) < 1e-6, f"sum@{tag}")
    row(f"sum@{tag}", f"reduce sum contig f64 N={n}",
        best_ms(lambda: np.sum(a), iters, warm, rounds))

    # copy (real ufunc nditer)
    np.positive(a, out=o)
    check(o[n - 1] == a[n - 1], f"copy@{tag}")
    row(f"copy@{tag}", f"unary copy contig f64 N={n}",
        best_ms(lambda: np.positive(a, out=o), iters, warm, rounds))

    # sadd
    np.add(sa, sb, out=so)
    check(so[n - 1] == sa[n - 1] + sb[n - 1], f"sadd@{tag}")
    row(f"sadd@{tag}", f"strided add a[::2]+b[::2] f64 N={n}",
        best_ms(lambda: np.add(sa, sb, out=so), iters, warm, rounds))

    # bcast
    np.add(a, b1, out=o)
    check(o[n - 1] == a[n - 1] + 3.0, f"bcast@{tag}")
    row(f"bcast@{tag}", f"broadcast add a+b1(1) f64 N={n}",
        best_ms(lambda: np.add(a, b1, out=o), iters, warm, rounds))

print("-" * 86)
print("ALL CORRECTNESS CHECKS PASS" if fails == 0 else f"{fails} CORRECTNESS FAILURES")

File: benchmark/poc/npyiter_sizesweep_bench.cs
#:project K:/source/NumSharp/src/NumSharp.Core/NumSharp.Core.csproj
#:property AssemblyName=NumSharp.DotNetRunScript
#:property PublishAot=false
#:property AllowUnsafeBlocks=true
// =============================================================================
// npyiter_sizesweep_bench.cs — the SAME six NpyIter operation families measured
// across four element-count tiers: scalar (1), 1K, 100K, 1M. Companion:
// npyiter_sizesweep_bench.py (NumPy side, identical ids).
//
// Mirrors the official benchmark report's per-size-tier layout, but for the
// iterator core. Each kernel is matched to NumPy's loop family so the measured
// time is dominated by what the ITERATOR pays (construction, broadcast/order
// resolution, per-chunk dispatch, iternext, reduction seed) at each size:
//   add   — contiguous binary, V256        vs np.add(a, b, out=o)
//   sqrt  — contiguous unary, V256 Sqrt     vs np.sqrt(a, out=o)
//   sum   — full reduction, 4-acc V256      vs np.sum(a)
//   copy  — contiguous copy, memcpy chunk   vs np.positive(a, out=o)  [real ufunc nditer]
//   sadd  — strided binary a[::2]+b[::2]     vs np.add(a2[::2], b2[::2], out=o)
//   bcast — stride-0 binary a + b1           vs np.add(a, b1, out=o)
//
// Run ONLY with:  dotnet run -c Release - < benchmark/poc/npyiter_sizesweep_bench.cs
// =============================================================================
using System.Diagnostics;
using System.Runtime.CompilerServices;
using System.Runtime.Intrinsics;
using NumSharp;
using NumSharp.Backends;
using NumSharp.Backends.Iteration;

var dbgScript = Attribute.GetCustomAttribute(typeof(Kern).Assembly, typeof(DebuggableAttribute)) as DebuggableAttribute;
var dbgCore = Attribute.GetCustomAttribute(typeof(np).Assembly, typeof(DebuggableAttribute)) as DebuggableAttribute;
if ((dbgScript?.IsJITOptimizerDisabled ?? false) || (dbgCore?.IsJITOptimizerDisabled ?? false))
{
    Console.WriteLine("FATAL: Debug-JITted assemblies (script or NumSharp.Core) — numbers would be INVALID.");
    Console.WriteLine("Run:   dotnet run -c Release - < benchmark/poc/npyiter_sizesweep_bench.cs");
    return;
}

int fails = 0;
void Check(bool ok, string what) { if (!ok) { fails++; Console.WriteLine($"  CORRECTNESS FAIL: {what}"); } }

double BestMs(Action body, int iters, int warm, int rounds = 5)
{
    for (int i = 0; i < warm; i++) body();
    double best = double.MaxValue;
    for (int r = 0; r < rounds; r++)
    {
        var sw = Stopwatch.StartNew();
        for (int i = 0; i < iters; i++) body();
        sw.Stop();
        best = Math.Min(best, sw.Elapsed.TotalMilliseconds / iters);
    }
    return best;
}

void Row(string id, string label, double ms)
{
    string val = ms >= 1.0 ? $"{ms,10:F3} ms" : ms >= 0.001 ? $"{ms * 1000,10:F2} us" : $"{ms * 1e6,10:F1} ns";
    Console.WriteLine($"{id,-12} {label,-46} {val}");
}

// iters/warm tuned per size so each (aspect,size) measures for a comparable wall time
(int iters, int warm) Pick(int n) =>
    n <= 1 ? (400_000, 40_000) :
    n <= 1_000 ? (200_000, 25_000) :
    n <= 100_000 ? (4_000, 600) :
    (400, 80);

var RO_WO = new[] { NpyIterPerOpFlags.READONLY, NpyIterPerOpFlags.WRITEONLY };
var RO_RO_WO = new[] { NpyIterPerOpFlags.READONLY, NpyIterPerOpFlags.READONLY, NpyIterPerOpFlags.WRITEONLY };
const NPY_ORDER K = NPY_ORDER.NPY_KEEPORDER;
const NPY_CASTING SAFE = NPY_CASTING.NPY_SAFE_CASTING;
const NpyIterGlobalFlags EXL = NpyIterGlobalFlags.EXTERNAL_LOOP;

Console.WriteLine($"NumSharp NpyIter size-sweep — {Environment.ProcessorCount} cores, V256={Vector256.IsHardwareAccelerated}");
Console.WriteLine($"{"id",-12} {"aspect",-46} {"per call",13}");
Console.WriteLine(new string('-', 86));

var SIZES = new (string tag, int n)[] { ("1", 1), ("1K", 1_000), ("100K", 100_000), ("1M", 1_000_000) };

unsafe
{
    foreach (var (tag, n) in SIZES)
    {
        var (iters, warm) = Pick(n);
        int rounds = n >= 100_000 ? 7 : 5;

        var a = (np.arange(n).astype(np.float64) % 97.0) + 1.0;
        var b = (np.arange(n).astype(np.float64) % 31.0) + 2.0;
        var o = np.empty(new Shape(n), np.float64);
        var b1 = NDArray.Scalar(3.0, NPTypeCode.Double).reshape(1);   // shape (1) -> broadcasts stride-0
        var a2 = (np.arange(2 * n).astype(np.float64) % 53.0) + 1.0;
        var b2 = (np.arange(2 * n).astype(np.float64) % 17.0) + 1.0;
        var sa = a2["::2"];
        var sb = b2["::2"];
        var so = np.empty(new Shape(n), np.float64);

        var add3 = new[] { a, b, o };
        var copy2 = new[] { a, o };
        var bc3 = new[] { a, b1, o };
        var sadd3 = new[] { sa, sb, so };

        // add — contiguous binary V256
        np.add(a, b, o);
        Check(o.GetDouble(n - 1) == a.GetDouble(n - 1) + b.GetDouble(n - 1), $"add@{tag}");
        Row($"add@{tag}", $"binary add contig f64 N={n}",
            BestMs(() => { using var it = NpyIterRef.MultiNew(3, add3, EXL, K, SAFE, RO_RO_WO); it.ForEach(Kern.AddF64); }, iters, warm, rounds));

        // sqrt — contiguous unary V256 Sqrt
        Row($"sqrt@{tag}", $"unary sqrt contig f64 N={n}",
            BestMs(() => { using var it = NpyIterRef.MultiNew(2, copy2, EXL, K, SAFE, RO_WO); it.ForEach(Kern.SqrtF64); }, iters, warm, rounds));
        Check(Math.Abs(o.GetDouble(n - 1) - Math.Sqrt(a.GetDouble(n - 1))) < 1e-9, $"sqrt@{tag}");

        // sum — full reduction
        double sum = 0;
        Row($"sum@{tag}", $"reduce sum contig f64 N={n}",
            BestMs(() => { using var it = NpyIterRef.New(a, EXL); sum = it.ExecuteReducing<SumLoop, double>(default, 0.0); }, iters, warm, rounds));
        Check(Math.Abs(sum - (double)np.sum(a)) / Math.Max(1.0, Math.Abs((double)np.sum(a))) < 1e-9, $"sum@{tag}");

        // copy — contiguous copy (real ufunc nditer on NumPy side = np.positive)
        Row($"copy@{tag}", $"unary copy contig f64 N={n}",
            BestMs(() => { using var it = NpyIterRef.MultiNew(2, copy2, EXL, K, SAFE, RO_WO); it.ForEach(Kern.CopyF64); }, iters, warm, rounds));
        Check(o.GetDouble(n - 1) == a.GetDouble(n - 1), $"copy@{tag}");

        // sadd — strided binary a[::2] + b[::2]
        np.add(sa, sb, so);
        Check(so.GetDouble(n - 1) == sa.GetDouble(n - 1) + sb.GetDouble(n - 1), $"sadd@{tag}");
        Row($"sadd@{tag}", $"strided add a[::2]+b[::2] f64 N={n}",
            BestMs(() => { using var it = NpyIterRef.MultiNew(3, sadd3, EXL, K, SAFE, RO_RO_WO); it.ForEach(Kern.AddF64); }, iters, warm, rounds));

        // bcast — stride-0 binary a + b1
        Row($"bcast@{tag}", $"broadcast add a+b1(1) f64 N={n}",
            BestMs(() => { using var it = NpyIterRef.MultiNew(3, bc3, EXL, K, SAFE, RO_RO_WO); it.ForEach(Kern.AddF64); }, iters, warm, rounds));
        Check(o.GetDouble(n - 1) == a.GetDouble(n - 1) + 3.0, $"bcast@{tag}");

        GC.Collect(); GC.WaitForPendingFinalizers();
    }
}

Console.WriteLine(new string('-', 86));
Console.WriteLine(fails == 0 ? "ALL CORRECTNESS CHECKS PASS" : $"{fails} CORRECTNESS FAILURES");
Console.Error.WriteLine("[done]");

static unsafe class Kern
{
    [MethodImpl(MethodImplOptions.AggressiveOptimization)]
    public static void CopyF64(void** dp, long* st, long count, void* aux)
    {
        byte* ps = (byte*)dp[0];
        byte* po = (byte*)dp[1];
        long ss = st[0], so = st[1];
        if (ss == 8 && so == 8) { Buffer.MemoryCopy(ps, po, count * 8, count * 8); return; }
        for (long i = 0; i < count; i++) { *(double*)po = *(double*)ps; ps += ss; po += so; }
    }

    [MethodImpl(MethodImplOptions.AggressiveOptimization)]
    public static void SqrtF64(void** dp, long* st, long count, void* aux)
    {
        byte* ps = (byte*)dp[0];
        byte* po = (byte*)dp[1];
        long ss = st[0], so = st[1];
        long i = 0;
        if (ss == 8 && so == 8)
        {
            for (; i + 4 <= count; i += 4)
            {
                Vector256.Store(Vector256.Sqrt(Vector256.Load((double*)ps)), (double*)po);
                ps += 32; po += 32;
            }
        }
        for (; i < count; i++) { *(double*)po = Math.Sqrt(*(double*)ps); ps += ss; po += so; }
    }

    [MethodImpl(MethodImplOptions.AggressiveOptimization)]
    public static void AddF64(void** dp, long* st, long count, void* aux)
    {
        byte* pa = (byte*)dp[0];
        byte* pb = (byte*)dp[1];
        byte* po = (byte*)dp[2];
        long sa = st[0], sb = st[1], so = st[2];
        long i = 0;
        if (so == 8)
        {
            if (sa == 8 && sb == 8)
            {
                for (; i + 8 <= count; i += 8)
                {
                    Vector256.Store(Vector256.Load((double*)pa) + Vector256.Load((double*)pb), (double*)po);
                    Vector256.Store(Vector256.Load((double*)(pa + 32)) + Vector256.Load((double*)(pb + 32)), (double*)(po + 32));
                    pa += 64; pb += 64; po += 64;
                }
            }
            else if (sa == 8 && sb == 0)
            {
                var vb = Vector256.Create(*(double*)pb);
                for (; i + 4 <= count; i += 4)
                {
                    Vector256.Store(Vector256.Load((double*)pa) + vb, (double*)po);
                    pa += 32; po += 32;
                }
            }
            else if (sa == 0 && sb == 8)
            {
                var va = Vector256.Create(*(double*)pa);
                for (; i + 4 <= count; i += 4)
                {
                    Vector256.Store(va + Vector256.Load((double*)pb), (double*)po);
                    pb += 32; po += 32;
                }
            }
        }
        for (; i < count; i++)
        {
            *(double*)po = *(double*)pa + *(double*)pb;
            pa += sa; pb += sb; po += so;
        }
    }
}

readonly struct SumLoop : INpyReducingInnerLoop<double>
{
    [MethodImpl(MethodImplOptions.AggressiveOptimization)]
    public unsafe bool Execute(void** dp, long* st, long count, ref double acc)
    {
        byte* ps = (byte*)dp[0];
        long ss = st[0];
        long i = 0;
        double s = 0;
        if (ss == 8)
        {
            var a0 = Vector256<double>.Zero;
            var a1 = Vector256<double>.Zero;
            var a2 = Vector256<double>.Zero;
            var a3 = Vector256<double>.Zero;
            for (; i + 16 <= count; i += 16)
            {
                a0 += Vector256.Load((double*)ps);
                a1 += Vector256.Load((double*)(ps + 32));
                a2 += Vector256.Load((double*)(ps + 64));
                a3 += Vector256.Load((double*)(ps + 96));
                ps += 128;
            }
            s = Vector256.Sum((a0 + a1) + (a2 + a3));
        }
        else
        {
            double s0 = 0, s1 = 0, s2 = 0, s3 = 0;
            for (; i + 4 <= count; i += 4)
            {
                s0 += *(double*)ps;
                s1 += *(double*)(ps + ss);
                s2 += *(double*)(ps + 2 * ss);
                s3 += *(double*)(ps + 3 * ss);
                ps += 4 * ss;
            }
            s = (s0 + s1) + (s2 + s3);
        }
        for (; i < count; i++) { s += *(double*)ps; ps += ss; }
        acc += s;
        return true;
    }
}
